src/Result.mo

```diff
@@ -85,53 +134,100 @@ public func fromOption<R, E>(x : ?R, err : E) : Result<R, E> {
   }
 };
 
-/// Maps the `Ok` type/value from the optional value, or else use the given error value.
-/// (Deprecate?)
-public func fromSomeMap<R1, R2, E>(x:?R1, f:R1->R2, err:E):Result<R2,E> {
-  switch x {
-    case (? x) {#ok (f x)};
-    case null {#err err};
+/// Create an option from a result, turning all #err into `null`.
+/// ```
+/// fromOption(#ok(x)) = ?x
+/// fromOption(#err(e)) = null
+/// ```
+public func toOption<R, E>(r : Result<R, E>) : ?R {
+  switch r {
+    case (#ok x) {?x};
+    case (#err _) {null};
   }
 };
 
-/// asserts that the option is Some(_) form.
-public func fromSome<Ok>(o:?Ok):Result<Ok,None> {
-  switch(o) {
-    case (?o) (#ok o);
-    case _ P.unreachable();
+/// Applies a function to a successful value, but discards the result. Use
+/// `iterate` if you're only interested in the side effect `f` produces.
+///
+/// ```
+/// var counter : Nat = 0;
+/// iterate<Nat, Text>(#ok(5), func (x : Nat) { counter += x });
+/// assert(counter == 5);
+/// iterate<Nat, Text>(#err("Wrong"), func (x : Nat) { counter += x });
+/// assert(counter == 5);
+/// ```
+public func iterate<Ok, Err>(res : Result<Ok, Err>, f : Ok -> ()) {
+  switch res {
+    case (#ok ok) f(ok);
+    case _ ();
   }
 };
 
-/// a result that consists of an array of Ok results from an array of results, or the first error in the result array, if any.
-public func toArrayOk<R,E>(x:[Result<R,E>]) : Result<[R],E> {
-  // return early with the first Err result, if any
-  for (i in x.keys()) {
-    switch (x[i]) {
-      case (#err e) { return #err(e) };
-      case (#ok _) { };
+/// Maps a Result-returning function over an Array and returns either
+/// the first error or an array of successful values.
+///
+/// ```motoko
+/// func makeNatural(x : Int) : Result<Nat, Text> =
+///   if (x >= 0) {
+///     #ok(Int.abs(x))
+///   } else {
+///     #err(Int.toText(x) # " is not a natural number.")
+///   };
+///
+/// traverseArray([0, 1, 2], makeNatural) = #ok([0, 1, 2]);
+/// traverseArray([-1, 0, 1], makeNatural) = #err("-1 is not a natural number.");
+/// ```
+public func traverseArray<A, R, E>(xs : [A], f : A -> Result<R, E>) : Result<[R], E> {
```

Comment:
Do we want this for mutable arrays too?

Reply:
Since this is type changing it couldn't actually mutate the Array. You mean a version that avoids the freeze the user would have to do otherwise?